bcftools stats module: more metrics and depth plot

chapmanb opened the pull request:
Phil;
I've been working on replacing some custom code/metrics with bcftools stats output, and this updates the module to provide some additional useful metrics and a depth plot of variants.

- Add heterozygote/homozygote counts to general statistics.
- Improves metrics naming in general statistics, emphasizing variant
  counts of interest.
- Avoid normalizing to million variants, since that is only useful for
  whole genome runs.
- Add plot of variants by depth for assessing coverage of variants

ewels commented:
Hi @chapmanb,

Many thanks for this, looks great! Are you happy to wait until I'm back from vacation in mid Feb for a review and merge, or would you like it done sooner? Happy to try to do it on my phone if so (but it'll be less thorough as I can't view the output or test).

Phil

chapmanb commented:
Phil -- absolutely I think it is your duty to review my PR on your phone in the middle of vacation. Seriously, have an awesome holiday and we can pick this back up when you're back. Hope you have a great time.

ewels commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -112,45 +128,56 @@ def parse_bcftools_stats(self):
+            # Make line graph of variants per depth
+            if len(depth_data) > 0:
+                import pprint
+                pprint.pprint(depth_data)
```
Looks like you have some debugging lines left in here..

ewels commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -112,45 +128,56 @@ def parse_bcftools_stats(self):
+                           'ylab': 'Fraction of sites (%)', 'xlab': 'Variant depth',
+                           'ymin': 0, 'ymax': 100, 'categories': True}
+                self.sections.append({'name': 'Variant depths',
+                                      'anchor': 'bcftools-stats',
```
This anchor needs to be unique and different to the plot above, otherwise the navigation links on the left of the report don't work.

I should probably implement a check for this in the code..

ewels commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -112,45 +128,56 @@ def parse_bcftools_stats(self):
+                           'ymin': 0, 'ymax': 100, 'categories': True}
+                self.sections.append({'name': 'Variant depths',
+                                      'anchor': 'bcftools-stats',
+                                      'content': linegraph.plot(depth_data, pconfig)})
```
Would be nice to add a line of text before the plot to explain what it's showing if possible..
```python
'content': '<p>This plot shows..</p>' + linegraph.plot(depth_data, pconfig)
```

ewels commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -112,45 +128,56 @@ def parse_bcftools_stats(self):
+                pprint.pprint(depth_data)
+                pconfig = {'id': 'bcftools_stats_depth', 'title': 'Variant depths',
+                           'ylab': 'Fraction of sites (%)', 'xlab': 'Variant depth',
+                           'ymin': 0, 'ymax': 100, 'categories': True}
```
Any reason for setting `'categories': True` here? Seems to be a numeric axis?

ewels replied:
Ah ok, just had a play - if I try to make `bin_name` a float when collecting the data I get an error because of one final one called `>500`. Not sure that there's really a way to get around this nicely so best as you have it I guess (though plot isn't as pretty).

ewels commented:
One other thing - I don't use bcftools myself so don't really have any view on which columns should be in the General Stats table, but it might be good to get feedback from others who use this module. Ping @barrymoore, @vladsaveliev, @darogan and @brentp...

Here are the proposed changes - going from this:
![image](https://cloud.githubusercontent.com/assets/465550/23133411/3e04d472-f791-11e6-93e5-f4f5a2e80a3c.png)

To this:
![image](https://cloud.githubusercontent.com/assets/465550/23133420/453cc510-f791-11e6-938e-130caa87a156.png)

Any comments from others?

ewels commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -112,45 +128,56 @@ def parse_bcftools_stats(self):
-            'min': 0,
-            'format': '{:.2f}',
-            'modify': lambda x: x / 1000000
+        stats_headers['number_of_records'] = {
```
Generally, I try to keep numbers in the General Stats under 1000 - hence the previous columns had the `'modify': lambda x: x / 1000000` to display the figures in millions.

My reasoning is that it's pretty difficult to spot the difference between 5522770 and 552277 when scanning a report quickly. The difference between 5.5 and 0.5 is easy though.

chapmanb replied:
Thanks, this makes good sense but the major reason I changed this is that the million variants representation is only useful for whole genome data. For exome or panels it ends up being a small decimal point which was confusing -- I got questions like why is 10,000 variants represented as 0.01M which is pretty fair. Maybe formatting with commas would solve both problems? I'm not sure if there's a straightforward way to do this within MultiQC now.

ewels replied:
Ah ok, this is a good point. Fine to leave it then I guess - could check the max value for all samples, but I guess it would be pretty confusing if it varied between reports.

I try to avoid commas where possible because of the German / Swedish / some-other-European-countries habit of using commas to denote decimal places (an even more annoying problem when parsing data). But I could make it use the same formatting as HighCharts with spaces (`5 522 770`). Already have the back end to customise this too: http://multiqc.info/docs/#number-formatting

ewels replied:
See #406

chapmanb commented:
Phil;
Thanks so much for the review, I addressed the issues you identified and commented on the one I'm not sure about. Let me know if you want to adjust anything else at all.

ewels commented:
Thanks Brad! I'll leave this open for a bit longer in case anyone else wants to comment on the change in General Stats fields, but will merge in a day or two if no-one complains :)

ewels commented:
Ok, that should have been enough time for people to complain 😀

amilesj commented:
It would great to add in a line plot of Variant Qualities, just like the one in the Whole-Genome Seq example on the MultiQC homepage that plots SNP quantity as a function of quality score, except using data from bcftools stats. 

PS I am newer to GitHub, so let me know if I should post this request elsewhere!

ewels commented:
Hi @amilesj,

That sounds doable - please can you open a new issue with that suggestion and it’ll be added to the list..

Phil

amilesj commented:
Thanks, will do!
